### PR TITLE
[RuntimeAsync] Fix a couple of cases where x86 would overwrite ecx on return

### PR DESCRIPTION
**Pull request by jakobbotsch**
shared-generics was not working on x86 because the returned continuation was being overwritten in one particular case. This fixes the issue.

### CODE REVIEW

**Comment by jakobbotsch** [comment]
PTAL @VSadov

File: src/coreclr/jit/codegenxarch.cpp [hunk 1]
```diff
@@ -10000,19 +10000,19 @@ void CodeGen::genFnEpilog(BasicBlock* block)
...
 #ifdef TARGET_X86
-            /* Add 'compiler->compLclFrameSize' to ESP */
-            /* Use pop ECX to increment ESP by 4, unless compiler->compJmpOpUsed is true */
+            // Add 'compiler->compLclFrameSize' to ESP. Use "pop ECX" for that, except in cases
+            // where ECX may contain some state.
 
-            if ((frameSize == TARGET_POINTER_SIZE) && !compiler->compJmpOpUsed)
+            if ((frameSize == TARGET_POINTER_SIZE) && !compiler->compJmpOpUsed && !compiler->compIsAsync2())
             {
                 inst_RV(INS_pop, REG_ECX, TYP_I_IMPL);
```
Comment (VSadov) [inline src/coreclr/jit/codegenxarch.cpp:10008]: If ECX is just a "some volatile register that we can trash", can we pop EDX instead?

Comment (jakobbotsch) [reply src/coreclr/jit/codegenxarch.cpp:10008]: `edx` is used to return longs on x86, so we can't easily switch to pop into that either.

Comment (VSadov) [reply src/coreclr/jit/codegenxarch.cpp:10008]: and if the goal is to have a smaller opcode - can x64 do "pop edx" ?  (just curious)

Comment (VSadov) [reply src/coreclr/jit/codegenxarch.cpp:10008]: I guess using pop on x64 may not fit some expected epilogue pattern or smth like that.

Comment (VSadov) [reply src/coreclr/jit/codegenxarch.cpp:10008]: Anyways, here we just want to fix runtime async issue on x86 - for that the fix looks good.

**Review by VSadov — APPROVED** [review]
LGTM!!

File: src/coreclr/jit/codegenxarch.cpp (hunk 1 above)
Comment (jakobbotsch) [reply src/coreclr/jit/codegenxarch.cpp:10008]: It does seem like we could optimize some epilog sequences in this way.
With Intel APX we are getting push2/pop2 instructions, and we are planning to look into prolog/epilog optimizations with those, so we'll probably look into it there.

**Pull request merged**